File: Difficulty: Medium/Maximize Number of 1's/maximize-number-of-1s.cpp
class Solution {
  public:
    int maxOnes(vector<int>& arr, int k) {
        int n = arr.size();
        int start = 0, end = 0;
        int maxLen = 0;
        
        while(end < n) {
            if(arr[end] == 1) {
                maxLen = max(maxLen, end - start + 1);
                end++;
            } else { // if(arr[end] == 0)
                if(k > 0) { // flip
                    k--;
                    maxLen = max(maxLen, end - start + 1);
                    end++;
                } else { // flip oldest flipped 0 to 1, and continue
                    while(start <= end && arr[start] != 0)
                        start++;
                    
                    start++;
                    
                    maxLen = max(maxLen, end - start + 1);
                    end++;
                }
            }
        }
        
        return maxLen;
    }
};

File: Difficulty: Medium/Maximize Number of 1's/maximize-number-of-1s.java
class Solution {
    public int maxOnes(int arr[], int k) {
        int n = arr.length, s = 0, e = 0, cnt = 0, ans = 0; 
        
        while(e < n) {
            if(arr[e] == 0) cnt++;
            while(cnt > k) {
                if(arr[s] == 0) cnt--;
                s++;
            }
            ans = Math.max(ans, e - s + 1);
            e++;
        }
        
        return ans;
    }
}